### PR TITLE
PyCOCOTools metric suite

### PR DESCRIPTION
**Pull request by LukeWood**
# What does this PR do?

Following my benchmarks in https://github.com/keras-team/keras-cv/pull/1515 I propose introducing a metric suite, still relying on PyCOCOTools.

This PR removes the in-graph COCO metrics in favor of a `PyMetric` flavored approach.  Originally I suspected that the implementation of PyCOCOTools was faster due to numpy stateful updates being faster, but it turns out that when we naively use numpy we receive even worse performance.  PyCOCOTools relies on custom C++ code, that we could eventually port to a KerasCV custom op, but we would need to further invest in infrastructure to make this maintainable.

This is unfortunately not the optimal solution to the "Object detection metrics" problem.  The best solution to "Object detection metrics" in my opinion is a suite of configurable COCO metrics:

i.e.
```
BoxRecall(area_range=(0, 100))
```

While I believe we compromise a bit on the API here, I still think that:
```
retinanet.compile(metrics=[keras_cv.metrics.BoxMetricSuite('xywh'))
```
Is a pretty decent API.  The biggest issue painpoint of PyCOCOTools is the print statements - this PR includes a context manager to temporarily print statement.

---

## Pros:

- allows us to move towards shipping a "Golden" OD workflow without further extreme investment into coco metric optimization
- efficient implementation
- guaranteed compatibility with any sort of publication --- everyone uses PyCOCOTools.
- great runtime due to PyCOCOTools C++ implementation of compute_iou()

225431775-f390b5e5-d590-4b43-98a0-af1d4b9d836b

---

## Cons:

1.) In this PR you will also find some really ugly code.  The unfortunate reality is that `tf.py_function` has some API shortcomings... namely:
- you can only input raw tensors (no dictionaries)
- you can only return a single tensor (no lists)

These restrictions make the use of @ianstenbit 's original `tf.keras.metrics.PyMetric` class unusable for this use case.  In this PR I've included a "hack" to make this work for our object detection use case, but its really not a good coding pattern.

2.) you can't customize the metrics.  They're hard coded - the meaning of "large" will always mean bigger than 96x96.

---

## Reviewing

Personally, I've struggled a great deal to optimize these metrics.  It is no small feat- and as such I believe this to be the prudent path forward.  In addition to this PR, I will produce two benchmarks: one based on https://github.com/keras-team/keras-cv/pull/1515 for recall, and another for mean average precision.  Inside of these benchmarks we can store the old implementations, and open this up as an "optimization challenge" to the community.

We've been working on COCO metrics for a long time - why didn't we just do this to start.  Honestly, the concept of a PYMetric is pretty recent.  If this existed 2 years ago and I was aware of the difficulty of properly implementing COCO metrics, I'd have advocated in favor of doing this from the start.

Before moving forwards with this direction, I'd like opinions from @ianstenbit @fchollet and @jbischof on the API, implementation, and the "PyCOCOTools" dependency.  Thanks for reading the long description.

Community members, feel free to comment - how do you feel about this API?

### CODE REVIEW

File: keras_cv/metrics/object_detection/object_detection_metric_suite.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+        box_loss='smoothl1',
+        optimizer=tf.optimizers.SGD(global_clipnorm=10.0),
+        jit_compile=False,
+        metrics=[keras_cv.metrics.ObjectDetectionMetricSuite('xywh')]
+    )
+    model.fit(images, labels)
+    ```
+    """
+
+    def __init__(self, bounding_box_format, eval_steps, name=None, **kwargs):
```
Comment (fchollet) [inline keras_cv/metrics/object_detection/object_detection_metric_suite.py:118]: `eval_steps` doesn't really convey the idea of recomputing the output every N step. In Keras "eval steps" kinda sounds like `validation_steps` which is the number of steps to run validation on in `fit()`.

Should probably reuse the arg names used in callbacks that do something every n steps, e.g. `save_freq`. Could be `recompute_freq`.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR. Looks reasonable I think!

**Comment by LukeWood** [comment]
@fchollet is there a way to disable the running-average on a  per step basis behavior in fit() on a per-metric level?   If not no problem, but it would be slightly better for these epoch-level metrics to just report the current exact value instead of an aggregation

File: keras_cv/metrics/object_detection/object_detection_metric_suite.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/metrics/object_detection/object_detection_metric_suite.py:118]: Done - thanks for the input on this API!

**Comment by LukeWood** [comment]
@fchollet new names for API + Parameters + updated docstring.  Let me know if you have any comments

File: keras_cv/metrics/object_detection/box_coco_metrics.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        box_loss='smoothl1',
+        optimizer=tf.optimizers.SGD(global_clipnorm=10.0),
+        jit_compile=False,
+        metrics=[keras_cv.metrics.BoxCOCOMetrics('xywh')]
+    )
+    model.fit(images, labels)
+    ```
+    """
+
+    def __init__(self, bounding_box_format, eval_steps, name=None, **kwargs):
```
Comment (fchollet) [inline keras_cv/metrics/object_detection/box_coco_metrics.py:123]: Update the arg name here and in the code please

File: keras_cv/metrics/object_detection/box_coco_metrics.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    "ARmax10",
+    "ARmax100",
+    "ARs",
+    "ARm",
+    "ARl",
+]
+
+
+class BoxCOCOMetrics(keras.metrics.Metric):
+    """BoxCOCOMetrics computes standard object deteciton metrics.
```
Comment (ianstenbit) [inline keras_cv/metrics/object_detection/box_coco_metrics.py:72]: nit: s/deteciton/detection

File: keras_cv/metrics/object_detection/box_coco_metrics.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+            values will allow for faster training times, while lower numbers
+            allow for higher numerical precision in metric reporting.
+
+    Usage:
+    `BoxCOCOMetrics()` can be used like any standard metric with any
+    KerasCV object detection model.  Inputs to `y_true` must be KerasCV bounding
+    box dictionaries, `{"classes": classes, "boxes": boxes}`, and `y_pred` must
+    follow the same format with an additional `confidence` key.
+
+    Using this metric suite alongside a model is trivial; simply provide it to
```
Comment (ianstenbit) [inline keras_cv/metrics/object_detection/box_coco_metrics.py:88]: I assume using this with our OD models breaks TPU compat. If so, probably worth mentioning that for TPU users should use the COCOCallback instead.

File: keras_cv/metrics/object_detection/box_coco_metrics.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+def _box_concat(b1, b2):
+    """Concatenates two bounding box batches together."""
+    if b1 is None:
+        return b2
+    if b2 is None:
+        return b1
+
+    result = {}
+    for key in ["boxes", "classes"]:
+        result[key] = tf.concat([b1[key], b2[key]], axis=0)
```
Comment (ianstenbit) [inline keras_cv/metrics/object_detection/box_coco_metrics.py:46]: I'd expect to see better performance if we keep all the intermediate tensors in a list and only concat them when we evaluate.

(This is what I saw when I tested PyMetric + COCO).

**Review by ianstenbit — CHANGES_REQUESTED** [review]
Thanks Luke!

File: keras_cv/metrics/object_detection/box_coco_metrics.py (hunk 4 above)
Comment (LukeWood) [reply keras_cv/metrics/object_detection/box_coco_metrics.py:88]: Right now my plan for TPU is actually to remove the COCO Callback and ship a generic PyMetric evaluator callback.  

We should not recommend use of the callback anymore, instead we should ship a generic API.

Comment (LukeWood) [reply keras_cv/metrics/object_detection/box_coco_metrics.py:88]: Updated with a blurb of insight on how to evaluate it right now.

File: keras_cv/metrics/object_detection/box_coco_metrics.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/metrics/object_detection/box_coco_metrics.py:123]: Totally missed this - thanks.

File: keras_cv/metrics/object_detection/box_coco_metrics.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+        self._eval_step_count = 0
+        self._cached_result = [0] * len(METRIC_NAMES)
+
+    def __new__(cls, *args, **kwargs):
+        obj = super(keras.metrics.Metric, cls).__new__(cls)
+
+        # Wrap the update_state function in a py_function and scope it to /cpu:0
+        obj_update_state = obj.update_state
+
+        def update_state_on_cpu(
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/metrics/object_detection/box_coco_metrics.py:138]: I think thats a great idea in the future.

Once we find a second use case for dicts I think that is the time to do this.

File: keras_cv/metrics/object_detection/box_coco_metrics.py (hunk 5 above)
Comment (LukeWood) [reply keras_cv/metrics/object_detection/box_coco_metrics.py]: Done - thanks for the pointer here.  tf.concat() in eager can be pretty expensive so makes sense to me as the latter approach only runs every eval_steps() epochs.

**Review by fchollet — APPROVED** [review]
LGTM

**Comment by LukeWood** [comment]
Thank you @ianstenbit and @fchollet for the reviews and making sure we can launch this!  I am _beyond_ happy to have something we feel good about for COCO metric evaluation.
